File: app/src/main/java/com/dumuzeyn/mp3player/MetadataValidator.java
package com.dumuzeyn.mp3player;

final class MetadataValidator {
    static final int MAX_TEXT_LENGTH = 300;

    private MetadataValidator() {
    }

    static String cleanText(String value) {
        if (value == null) {
            return "";
        }
        String cleaned = value.replaceAll("[\\p{Cntrl}&&[^\\n\\t]]", "")
                .replace('\n', ' ').replace('\t', ' ').replaceAll("\\s+", " ").trim();
        return cleaned.length() <= MAX_TEXT_LENGTH
                ? cleaned : cleaned.substring(0, MAX_TEXT_LENGTH).trim();
    }

    static int year(String value) {
        return boundedNumber(value, 0, 3000);
    }

    static int trackNumber(String value) {
        return boundedNumber(firstPart(value), 0, 9999);
    }

    private static String firstPart(String value) {
        if (value == null) {
            return "";
        }
        int separator = value.indexOf('/');
        return separator < 0 ? value : value.substring(0, separator);
    }

    private static int boundedNumber(String value, int minimum, int maximum) {
        try {
            int parsed = Integer.parseInt(value == null ? "" : value.trim());
            return parsed >= minimum && parsed <= maximum ? parsed : 0;
        } catch (NumberFormatException error) {
            return 0;
        }
    }
}

File: app/src/main/java/com/dumuzeyn/mp3player/MetadataValidator.kt
package com.dumuzeyn.mp3player

object MetadataValidator {
    const val MAX_TEXT_LENGTH = 300
    private val controls = Regex("[\\p{Cntrl}&&[^\\n\\t]]")
    private val whitespace = Regex("\\s+")

    @JvmStatic
    fun cleanText(value: String?): String {
        val cleaned = value.orEmpty()
            .replace(controls, "")
            .replace('\n', ' ')
            .replace('\t', ' ')
            .replace(whitespace, " ")
            .trim()
        return if (cleaned.length <= MAX_TEXT_LENGTH) cleaned else cleaned
            .substring(0, MAX_TEXT_LENGTH)
            .trim()
    }

    @JvmStatic
    fun year(value: String?): Int = boundedNumber(value, 0, 3_000)

    @JvmStatic
    fun trackNumber(value: String?): Int = boundedNumber(value.orEmpty().substringBefore('/'), 0, 9_999)

    private fun boundedNumber(value: String?, minimum: Int, maximum: Int): Int =
        value?.trim()?.toIntOrNull()?.takeIf { it in minimum..maximum } ?: 0
}
